### DISCUSSION TITLE
Using st25r95 with ESP32

### DISCUSSION THREAD

**Discussion by alanblu** [stm32duino/Arduino_Core_STM32#2342] [Libraries]
Hello everyone,

I am wondering if there is any example of usin the card reader st25r95 with the ESP through SPI.

Regards
Alan

**Comment by fpistm** [comment]
Hi @alanblu 

did you try this library:
https://github.com/stm32duino/ST25R95
You can see how it is used thanks:
https://github.com/stm32duino/X-NUCLEO-NFC03A1/blob/main/examples/X_NUCLEO_NFC03A1_HelloWorld/X_NUCLEO_NFC03A1_HelloWorld.ino

Let us know if it works then we could set ESP as compatible with the library.
/cc @cparata

**Reply by alanblu** [reply]
Hi

I did try only the includes in an ino file:

#include "nfc_utils.h"
#include "rfal_nfc.h"
#include "rfal_rfst25r95.h"
#include "ndef_class.h"

void setup() {
  // put your setup code here, to run once:

}

void loop() {
  // put your main code here, to run repeatedly:

}

But I got the error that since is another architechture, thos libraries wont work:

WARNING: library STM32duino NFC-RFAL claims to run on stm32 architecture(s) and may be incompatible with your current board which runs on esp32 architecture(s).
WARNING: library STM32duino ST25R95 claims to run on stm32 architecture(s) and may be incompatible with your current board which runs on esp32 architecture(s).

specreg.h:40:13: error: expected unqualified-id before numeric constant
#define BR  4


rfal_nfcDep.h:339:22: note: in expansion of macro 'BR'
   uint8_t            BR;             /*!< Initiator's Bit Rates supported in Rx */
   
   exit status 1

Compilation error: exit status 1

**Reply by fpistm** [reply]
As @cparata said, ESP core define BR:
https://github.com/search?q=org%3Aespressif%20%2F%23define%20BR%5Cs%2B4%2F&type=code

But the library use BR as a member name:
https://github.com/stm32duino/NFC-RFAL/blob/cee190c24a8165e399821a4e0c51098cdd2eb227/src/rfal_nfcDep.h#L339

**Reply by alanblu** [reply]
After redifining BR it compiles but still after using rfalNfcInitialize(); I am getting error 8 (ERR_SYSTEM)

**Reply by cparata** [reply]
Hello @alanblu ,
unfortunately, I don't have an ESP board to reproduce your setup. What is your hardware setup? Do you have the possibility to use a logical analyzer to check the SPI signals? 
Best Regards,
Carlo

**Comment by cparata** [comment]
Hello @alanblu ,
it seems to me that some library defines are in conflict with some definitions in the ESP core.
Best Regards,
Carlo

**Comment by alanblu** [comment]
Thanks for the answer, I'm trying to use a configuration without irq_in,
not sure if that is possible

Carlo Parata ***@***.***> schrieb am Do., 16. Mai 2024, 18:47:

**Reply by cparata** [reply]
Hi @alanblu ,
unfortunately not. The IRQ_IN is mandatory for the correct usage of the NFC reader library.
Best Regards,
Carlo

**Comment by alanblu** [comment]
I added a while for the configuration and now the error is not an issue anymore:

`  
  while(err != ERR_NONE)
  {
     err = rfal_nfc.rfalNfcInitialize();
  }
  
  if (err == ERR_NONE) {
    discParam.compMode      = RFAL_COMPLIANCE_MODE_NFC;
    discParam.devLimit      = 1U;
    discParam.nfcfBR        = RFAL_BR_212;
    discParam.ap2pBR        = RFAL_BR_424;
    ST_MEMCPY(&discParam.nfcid3, NFCID3, sizeof(NFCID3));
    ST_MEMCPY(&discParam.GB, GB, sizeof(GB));
    discParam.GBLen         = sizeof(GB);

    discParam.notifyCb             = NULL;
    discParam.totalDuration        = 1000U;
    discParam.wakeupEnabled        = false;
    discParam.wakeupConfigDefault  = true;
    discParam.techs2Find           = (RFAL_NFC_POLL_TECH_A | RFAL_NFC_POLL_TECH_B | RFAL_NFC_POLL_TECH_F | RFAL_NFC_POLL_TECH_V | RFAL_NFC_POLL_TECH_ST25TB);

    state = DEMO_ST_START_DISCOVERY;
  }
`

But still when I call rfalNfcGetState() I alway get  RFAL_NFC_STATE_NOTINIT

**Reply by cparata** [reply]
Are you using the IRQ_IN pin now? Because it is mandatory for the correct usage of the library.
Best Regards,
Carlo

**Comment by alanblu** [comment]
Yes, I'm using that pin

Carlo Parata ***@***.***> schrieb am Mi., 22. Mai 2024, 16:35:

**Reply by cparata** [reply]
Hello @alanblu ,
if you get RFAL_NFC_STATE_NOTINIT state, it means that the "rfalInitialize" in the "rfal_rfst25r95.cpp" file returns an error. Could you put some prints inside this function, so we can understand where is the error? Do you have the chance to check if the SPI communication is working on your hardware setup?
Best Regards,
Carlo

**Reply by alanblu** [reply]
Sorry for the delay. I checked the SPI communication and it looks weird, I tried with another software with SPI communication and that looks good, but with the example software for arduino, it doesnt generate any SPI communication. Is there any example for ESP32?

**Reply by cparata** [reply]
Hello @alanblu ,
how did you connect the st25r95 to the ESP? Could you share your hardware setup?
Best Regards,
Carlo

**Reply by alanblu** [reply]
This is the connexion we are using:

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/4308583/6a88205b-e6b1-4b1b-8b7a-cfb225a9e76d)

and the connection to the microcontroller:

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/4308583/0f360d31-4ff3-48a9-82ae-fad2cc0892c3)

**Comment by cparata** [comment]
The schematics seem fine to me. Are you sure that you are using the correct SPI instance for ESP?
Best Regards,
Carlo

**Reply by alanblu** [reply]
It works with other project in vscode and the same board (no arduino), but I am not generating SPI signal with the arduino project.

Those are the pins:
#define SPI_MOSI 11
#define SPI_MISO 13
#define SPI_SCK 12
#define CS_PIN 10
#define LED_A_PIN 12
#define LED_B_PIN 11
#define LED_F_PIN 13
#define LED_V_PIN 4
#define IRQ_IN_PIN 8
#define IRQ_OUT_PIN 9
#define INTERFACE_PIN 16
#define USER_BTN 14

**Comment by cparata** [comment]
Are you at least able to see the SPI clock with a logical analyzer?

**Reply by alanblu** [reply]
not with the arduino software

**Reply by cparata** [reply]
We should check the SPI signals when the firmware does not work (with Arduino software). Otherwise, it's difficult to understand where is the issue.
Best Regards,
Carlo

**Reply by alanblu** [reply]
Hello, I finally managed to get some screenshot of the SPI bus:

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/4308583/bdc8f20d-c2d9-494e-ad60-a60a43ff08a9)

It kind of look weird

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/4308583/cac561ad-d08c-4a95-a2ba-531a1fd7812a)

Best Regards
Alan

**Comment by cparata** [comment]
Hello @alanblu ,
apparently it seems that the MCU does not start the SPI clock. This is really strange. Are you sure that the SPI clock is well connected to the MCU?
Best Regards,
Carlo

**Reply by alanblu** [reply]
I tested by switching all the pins with an Independent code and it is working

**Reply by alanblu** [reply]
Also SPI is working, I am getting 0x06 in Miso

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/4308583/242813fc-f38f-4c5d-8581-da464ab2ca43)

**Reply by cparata** [reply]
Ok, are you able to check where the code is stuck? Unfortunately, without your hardware setup is difficult for me to debug the issue.

**Reply by alanblu** [reply]
I had some issues with the pins, the ones in for the LEDs were the same for my some spi pins, now the bus is working but I am still getting ERR_SYSTEM after running rfalNfcInitialize()

This is the spi bus:

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/4308583/e5b2f8d4-4b30-4373-9a06-64ae80742667)

And you can see in this bus that IRQ_OUT is working:

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/4308583/92ea5b0d-2f1a-4ea2-a288-735f22023863)

**Reply by alanblu** [reply]
Hello, I still need some help here! 

Thanks in advance
Alan Blumenstein

**Reply by cparata** [reply]
Hi @alanblu ,
unfortunately we do not have ESP hardware to reproduce. Moreover library is not compatible with ESP core (you was obliged to perform a BR redefinition to get it compiling). So, we could not help more on this subject.
I'm sorry for that.
Best Regards,
Carlo
